File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/metadata/fetcher/Brave.java
package de.marmaro.krt.ffupdater.metadata.fetcher;

import java.net.URL;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.Callable;

import de.marmaro.krt.ffupdater.device.ABI;
import de.marmaro.krt.ffupdater.device.DeviceEnvironment;
import de.marmaro.krt.ffupdater.metadata.AvailableMetadata;
import de.marmaro.krt.ffupdater.metadata.ReleaseVersion;
import de.marmaro.krt.ffupdater.utils.ParamRuntimeException;

/**
 * https://github.com/mozilla-tw/FirefoxLite/releases
 * https://api.github.com/repos/mozilla-lockwise/lockwise-android/releases/latest
 */
class Brave implements Callable<AvailableMetadata> {
    private final GithubConsumer githubConsumer;
    private final GithubConsumer.Request request;
    private final DeviceEnvironment deviceEnvironment;

    Brave(GithubConsumer githubConsumer, DeviceEnvironment deviceEnvironment) {
        this.deviceEnvironment = deviceEnvironment;
        Objects.requireNonNull(githubConsumer);
        this.githubConsumer = githubConsumer;
        request = new GithubConsumer.Request()
                .setOwnerOfRepository("brave")
                .setRepositoryName("brave-browser")
                .setResultsPerPage(20)
                .setReleaseValidator(this::isReleaseValid);
    }

    private boolean isReleaseValid(GithubConsumer.Release release) {
        final String name = release.getName();
        if (name == null || !name.startsWith("Release ")) {
            return false;
        }
        if (release.isPrerelease()) {
            return false;
        }
        final List<GithubConsumer.Asset> assets = release.getAssets();
        if (assets == null) {
            return false;
        }
        for (GithubConsumer.Asset asset : assets) {
            if (asset.getName().endsWith(".apk")) {
                return true;
            }
        }
        return false;
    }

    @Override
    public AvailableMetadata call() {
        final GithubConsumer.GithubResult result = githubConsumer.consumeManyReleases(request);

        final URL downloadURL;
        final ABI abi = deviceEnvironment.getSupportedAbis().get(0);
        switch (abi) {
            case AARCH64:
                downloadURL = result.getUrls().get("BraveMonoarm64.apk");
                break;
            case ARM:
                downloadURL = result.getUrls().get("BraveMonoarm.apk");
                break;
            case X86:
                downloadURL = result.getUrls().get("BraveMonox86.apk");
                break;
            case X86_64:
                downloadURL = result.getUrls().get("BraveMonox64.apk");
                break;
            default:
                throw new ParamRuntimeException("unknown ABI %s - switch fallthrough", abi);
        }
        Objects.requireNonNull(downloadURL);

        final String versionName = result.getTagName().replace("v", "");
        return new AvailableMetadata(new ReleaseVersion(versionName), downloadURL);
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/impl/Brave.kt
package de.marmaro.krt.ffupdater.app.impl

import android.content.Context
import android.os.Build
import android.util.Log
import androidx.annotation.MainThread
import de.marmaro.krt.ffupdater.R
import de.marmaro.krt.ffupdater.app.App
import de.marmaro.krt.ffupdater.app.entity.DisplayCategory
import de.marmaro.krt.ffupdater.app.entity.LatestUpdate
import de.marmaro.krt.ffupdater.device.ABI.*
import de.marmaro.krt.ffupdater.device.DeviceAbiExtractor
import de.marmaro.krt.ffupdater.device.DeviceSdkTester
import de.marmaro.krt.ffupdater.network.exceptions.NetworkException
import de.marmaro.krt.ffupdater.network.file.CacheBehaviour
import de.marmaro.krt.ffupdater.network.github.GithubConsumer
import de.marmaro.krt.ffupdater.settings.DeviceSettingsHelper

/**
 * https://github.com/brave/brave-browser/releases
 * https://api.github.com/repos/brave/brave-browser/releases
 * https://www.apkmirror.com/apk/brave-software/brave-browser/
 */
class Brave : AppBase() {
    override val app: App = App.BRAVE
    override val packageName = "com.brave.browser"
    override val title = R.string.brave__title
    override val description = R.string.brave__description
    override val installationWarning = R.string.brave__warning
    override val downloadSource = "GitHub"
    override val icon = R.drawable.ic_logo_brave
    override val minApiLevel = Build.VERSION_CODES.M
    override val supportedAbis = ARM32_ARM64_X86_X64

    @Suppress("SpellCheckingInspection")
    override val signatureHash = "9c2db70513515fdbfbbc585b3edf3d7123d4dc67c94ffd306361c1d79bbf18ac"
    override val projectPage = "https://github.com/brave/brave-browser"
    override val displayCategory = DisplayCategory.BETTER_THAN_GOOGLE_CHROME

    @MainThread
    @Throws(NetworkException::class)
    override suspend fun findLatestUpdate(
        context: Context,
        cacheBehaviour: CacheBehaviour,
    ): LatestUpdate {
        val time = System.nanoTime()
        Log.d(LOG_TAG, "check for latest version")
        val fileName = findNameOfApkFile()
        val result = GithubConsumer.findLatestRelease(
            repository = GithubConsumer.REPOSITORY__BRAVE__BRAVE_BROWSER,
            resultsPerApiCall = GithubConsumer.RESULTS_PER_API_CALL__BRAVE_BROWSER,
            dontUseApiForLatestRelease = true,
            isValidRelease = { !it.isPreRelease && it.name.startsWith("Release v") },
            isSuitableAsset = { it.name == fileName },
            cacheBehaviour = cacheBehaviour,
        )
        val version = result.tagName.replace("v", "")
        Log.i(LOG_TAG, "found latest version $version after ${(System.nanoTime() - time) / 1000000}ms")
        return LatestUpdate(
            downloadUrl = result.url,
            version = version,
            publishDate = result.releaseDate,
            exactFileSizeBytesOfDownload = result.fileSizeBytes,
            fileHash = null,
        )
    }

    private fun findNameOfApkFile(): String {
        val strings = if (DeviceSdkTester.supportsAndroidNougat()) {
            DeviceAbiExtractor.StringsForAbi(
                armeabi_v7a = "BraveMonoarm.apk",
                arm64_v8a = "BraveMonoarm64.apk",
                x86 = "BraveMonox86.apk",
                x86_64 = "BraveMonox64.apk"
            )
        } else {
            DeviceAbiExtractor.StringsForAbi(
                armeabi_v7a = "Bravearm.apk",
                x86 = "Bravex86.apk",
            )
        }
        return DeviceAbiExtractor.findStringForBestAbi(strings, DeviceSettingsHelper.prefer32BitApks)
    }

    companion object {
        private const val LOG_TAG = "Brave"
    }
}